### PR TITLE
[htslib, htscodecs] New ports

### PR DESCRIPTION
**Pull request by teo-tsirpanis**
- [X] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [X] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [X] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [X] The versioning scheme in `vcpkg.json` matches what upstream says.
- [X] The license declaration in `vcpkg.json` matches what upstream says.
- [X] The installed as the "copyright" file matches what upstream says.
- [X] The source code of the component installed comes from an authoritative source.
- [X] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [X] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [X] Only one version is in the new port's versions file.
- [X] Only one version is added to each modified port's versions file.

### CODE REVIEW

File: ports/htslib/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,44 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO samtools/htslib
+    REF "${VERSION}"
+    SHA512 b9de3769db6153f66348c7c4ffbfc5ac7cd4a4d4450c9d1c5ea8fdd8f4f9d38d1d0ba5b4ac9c53f1a754d3985dc483fe22e76f93a8bbe8ae29ef3b98136e7d2e
+    HEAD_REF develop
+    PATCHES
+        0001-set-linkage-${VCPKG_LIBRARY_LINKAGE}.patch
```
Comment (dg0yt) [inline ports/htslib/portfile.cmake:8]: Do not use conditional patching.
Use a single patch which activates the right thing subject to configuration.

File: ports/htscodecs/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO samtools/htscodecs
+    REF "v${VERSION}"
+    SHA512 7533570b8b1cad0b9ed118170e5a9ff34fdde40b090f4ba3756f7899e2cd7230f8172425ecf6bd7b83b0b0b1a2a24f3d21795db7f0bc2c3add0a55342b970d1a
+    HEAD_REF master
+    PATCHES
+        0001-no-tests.patch # https://github.com/samtools/htscodecs/pull/120
+)
+
+set(FEATURE_OPTIONS)
```
Comment (dg0yt) [inline ports/htscodecs/portfile.cmake:11]: This works different than expected due to scope rules. Either set to `""`, or 
```suggestion
vcpkg_list(SET FEATURE_OPTIONS)
```

**Review by dg0yt — CHANGES_REQUESTED** [review]
The `has_${feature}` variables are unused AFAICS.

**Comment by jimwang118** [comment]
@teo-tsirpanis Thanks for your PR, Is work still being done for this PR?

**Comment by teo-tsirpanis** [comment]
I will look at the Android failures later.

File: ports/htslib/portfile.cmake (hunk 1 above)
Comment (teo-tsirpanis) [reply ports/htslib/portfile.cmake]: Done.

File: ports/htscodecs/portfile.cmake (hunk 2 above)
Comment (teo-tsirpanis) [reply ports/htscodecs/portfile.cmake:11]: Done.

**Comment by teo-tsirpanis** [comment]
I can't download the artifacts to check the Android failures.

**Comment by dg0yt** [comment]
> I can't download the artifacts to check the Android failures.

Can be downloaded as always.
~~~
ld.lld: error: unable to find library -lpthread
clang-17: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [Makefile:361: libhts.so] Error 1
make: *** Waiting for unfinished jobs....
~~~
This should probably be `-pthread`, for most modern systems.

**Comment by teo-tsirpanis** [comment]
Thanks @dg0yt! I finally found where the artifacts were located in AzDo.

**Comment by teo-tsirpanis** [comment]
CI is green.

**Comment by jimwang118** [comment]
All features passed with following triplet:

```
x64-linux
```

File: ports/htscodecs/vcpkg.json [hunk 3]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "htscodecs",
+  "version": "1.6.0",
+  "description": "Custom compression for CRAM and others.",
+  "homepage": "https://github.com/samtools/htscodecs",
+  "license": "MIT",
+  "supports": "!windows",
+  "dependencies": [
+    "pthreads"
+  ],
```
Comment (dg0yt) [inline ports/htscodecs/vcpkg.json:10]: IMO this doesn't add anything useful here. Some port depend on pthreads for windows, but this port does not support windows.

File: ports/htslib/vcpkg.json [hunk 4]
```diff
@@ -0,0 +1,33 @@
+{
+  "name": "htslib",
+  "version": "1.20",
+  "description": "C library for high-throughput sequencing data formats",
+  "homepage": "https://github.com/samtools/htslib",
+  "license": "MIT",
+  "supports": "!windows",
+  "dependencies": [
+    "htscodecs",
+    "pthreads",
```
Comment (dg0yt) [inline ports/htslib/vcpkg.json:10]:
```suggestion
```

**Review by dg0yt — COMMENTED** [review]
(Opinion)

File: ports/htslib/portfile.cmake [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    REF "${VERSION}"
+    SHA512 b9de3769db6153f66348c7c4ffbfc5ac7cd4a4d4450c9d1c5ea8fdd8f4f9d38d1d0ba5b4ac9c53f1a754d3985dc483fe22e76f93a8bbe8ae29ef3b98136e7d2e
+    HEAD_REF develop
+    PATCHES
+        0001-set-linkage.patch
+        0002-pthread-flag.patch
+        0003-no-tests.patch
+)
+
+SET(FEATURE_OPTIONS "")
```
Comment (dg0yt) [inline ports/htslib/portfile.cmake:13]: Nitpick:
```suggestion
set(FEATURE_OPTIONS "")
```

**Comment by teo-tsirpanis** [comment]
Feedback addressed.

**Comment by JavierMatosD** [comment]
I was unable to install these ports. I'm using the a code spaces instance with Ubuntu (I believe)

```
% vcpkg install htslib htscodecs

CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:127 (message):
    Command failed: /usr/bin/bash -c "V=1 ./../src/1.20-30e933fbb9.clean/configure  \"--with-external-htscodecs\" \"--disable-libcurl\" \"--disable-gcs\" \"--disable-s3\" \"--disable-plugins\" \"--disable-bz2\" \"--disable-lzma\" \"--without-libdeflate\" \"--disable-silent-rules\" \"--verbose\" \"--disable-shared\" \"--enable-static\" \"--prefix=/workspaces/vcpkg/installed/x64-linux/debug\" \"--bindir=\\${prefix}/../tools/htslib/debug/bin\" \"--sbindir=\\${prefix}/../tools/htslib/debug/sbin\" \"--libdir=\\${prefix}/lib\" \"--includedir=\\${prefix}/../include\" \"--datarootdir=\\${prefix}/share/htslib\""
    Working Directory: /workspaces/vcpkg/buildtrees/htslib/x64-linux-dbg
    Error code: 1
    See logs for more information:
      /workspaces/vcpkg/buildtrees/htslib/config-x64-linux-dbg-config.log
      /workspaces/vcpkg/buildtrees/htslib/config-x64-linux-dbg-out.log
      /workspaces/vcpkg/buildtrees/htslib/config-x64-linux-dbg-err.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_configure_make.cmake:866 (vcpkg_execute_required_process)
  ports/htslib/portfile.cmake:32 (vcpkg_configure_make)
  scripts/ports.cmake:192 (include)
```


/workspaces/vcpkg/buildtrees/htslib/config-x64-linux-dbg-err.log
```
configure: WARNING: unrecognized options: --disable-silent-rules, --disable-shared, --enable-static
configure: error: libhtscodecs development files not found: missing library

You asked to use an external htscodecs library, but do not have the
required header / library files.  You either need to supply these and
if necessary set CPPFLAGS and LDFLAGS so the compiler can find them;
or configure using --without-external-htscodecs to build the required
functions from the htscodecs submodule.

```

It looks like it can't find:

- htscodecs/htscodecs.h
- libhtscodecs.so

**Comment by teo-tsirpanis** [comment]
Weird. Do these files exist? Did htscodecs build correctly? Can you try installing only htslib?

**Comment by teo-tsirpanis** [comment]
Apparently when trying to link to `-lhtscodecs` it fails because it does not find pthread. 🤔

**Comment by teo-tsirpanis** [comment]
After the latest patch, installing succeeds on a codespace.

**Pull request merged**